### PR TITLE
[Spark][Kernel] Pass table identifier in Coordinated Commits

### PR DESCRIPTION
**Pull request by prakharjain09**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [x] Kernel
- [ ] Other (fill in here)

## Description

This PR adds TableIdentifier in the Coordinated Commit Interface.

There will be a separate change to pass the tableName reliably to the Commit Coordinator in delta-spark.

## How was this patch tested?

Existing UTs

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

**Comment by scovich** [comment]
Missing some newly added files by chance?
```
[error] /home/runner/work/delta/delta/storage/src/main/java/io/delta/storage/commit/CommitCoordinatorClient.java:73:1: cannot find symbol
[error]   symbol:   class TableIdentifier
[error]   location: interface io.delta.storage.commit.CommitCoordinatorClient
[error] TableIdentifier
[error] /home/runner/work/delta/delta/storage/src/main/java/io/delta/storage/commit/CommitCoordinatorClient.java:97:1: cannot find symbol
[error]   symbol:   class TableDescriptor
[error]   location: interface io.delta.storage.commit.CommitCoordinatorClient
[error] TableDescriptor
```

**Comment by prakharjain09** [comment]
> 
> Missing some newly added files by chance?
> 
> [error] /home/runner/work/delta/delta/storage/src/main/java/io/delta/storage/commit/CommitCoordinatorClient.java:73:1: cannot find symbol
> [error]   symbol:   class TableIdentifier
> [error]   location: interface io.delta.storage.commit.CommitCoordinatorClient
> [error] TableIdentifier
> [error] /home/runner/work/delta/delta/storage/src/main/java/io/delta/storage/commit/CommitCoordinatorClient.java:97:1: cannot find symbol
> [error]   symbol:   class TableDescriptor
> [error]   location: interface io.delta.storage.commit.CommitCoordinatorClient
> [error] TableDescriptor
>

Yes - fixed it.

**Review by scovich — COMMENTED** [review]
This PR seems to mix two things:
1. Introducing the `TableDescriptor` and plumbing it through all the various method signatures and call sites. Important, but very noisy.
2. Introducing the `TableIdentifier` and barely starting to plumb it in a couple of spots (which just pass a hard-wired empty option). 

Can we move the second logic to the PR that actually uses table identifiers in a meaningful way? Then this PR would be a lot easier to review, and probably that later PR as well.

**Comment by prakharjain09** [comment]
> Can we move the second part to the PR that actually plumbs through table identifiers in a meaningful way?

The changes to plumb TableIdentifier is a superset of the changes to plumb TableDescriptor. This PR focuses on the TableDescriptor part and together with it also takes care of plumbing TableIdentifier in those areas. Have left specific TODOs to fix the call-sites and pass a non-empty TableIdentifier which will be taken in next set of PRs.

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/engine/DefaultCommitCoordinatorClientHandler.java [hunk 1]
```diff
@@ -89,8 +91,11 @@ public Map<String, String> registerTable(
       long currentVersion,
       AbstractMetadata currentMetadata,
       AbstractProtocol currentProtocol) {
+    // TODO: Introduce table identifier concept in Table API in Kernel and plumb the
+    //  table identifier into `CommitCoordinatorClient` in all APIs.
     return commitCoordinatorClient.registerTable(
         new Path(logPath),
+        Optional.empty(),
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/engine/DefaultCommitCoordinatorClientHandler.java:98]:
```suggestion
        Optional.empty() /* table identfier */,
```

Comment (vkorukanti) [reply kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/engine/DefaultCommitCoordinatorClientHandler.java:98]: same in other places.

**Review by vkorukanti — APPROVED** [review]
Kernel changes LGTM

File: spark/src/test/scala/org/apache/spark/sql/delta/test/DeltaTestImplicits.scala [hunk 2]
```diff
@@ -91,6 +93,40 @@ object DeltaTestImplicits {
     }
   }
 
+  /** Helper class for working with [[TableCommitCoordinatorClient]] */
```
Comment (scovich) [inline spark/src/test/scala/org/apache/spark/sql/delta/test/DeltaTestImplicits.scala:96]: These helpers avoid having to retrofit zillions of call sites in path-based unit tests?

**Review by scovich — APPROVED** [review]
LGTM. Just double check some files that may import java Optional unnecessarily?

File: spark/src/test/scala/org/apache/spark/sql/delta/test/DeltaTestImplicits.scala (hunk 2 above)
Comment (prakharjain09) [reply spark/src/test/scala/org/apache/spark/sql/delta/test/DeltaTestImplicits.scala:96]: Yes - these are very useful.

**Pull request merged**